internal/charmstore/search.go

```diff
@@ -871,3 +833,42 @@ func createElasticSort(s SortParam) elasticsearch.Sort {
 	}
 	return sort
 }
+
+// searchFields maps the fields that will be requested in the Iter fields
+// parameter to the required elasticsearch source filter.
+var searchFields = func() map[string]string {
+	t := reflect.ValueOf(mongodoc.Entity{}).Type()
+	jsonFields := fieldNames(t, "json", nil)
+	mapping := make(map[string]string, len(jsonFields))
+	for i, bsonField := range fieldNames(t, "bson", strings.ToLower) {
+		mapping[bsonField] = jsonFields[i]
+	}
+	return mapping
+}()
+
+// fieldNames calculates the field names for documents derived from the
+// given type. A field name is added to the result for every exported
+// field in t. If the field has a tag with the given key then the name be
+// the first field in the tags value after it has been comma separated.
+// If there is no tag, or the name is empty then the name is derived by
+// calling the given function on the field name. If f is nil the field
+// name will be used without transformation.
+func fieldNames(t reflect.Type, key string, f func(string) string) []string {
```

Comment:
// Note: this does not support embedded fields, but we don't have any embedded
// structs in the types we care about yet so that's OK.

?